direct/core/src/main/java/cz/o2/proxima/direct/transaction/TransactionalOnlineAttributeWriter.java

```diff
@@ -227,6 +269,17 @@ private TransactionalOnlineAttributeWriter(
     // FIXME: this should be removed and resolved automatically as empty input
     // to be fixed in https://github.com/O2-Czech-Republic/proxima-platform/issues/216
     this.globalKeyAttributes = getAttributesWithGlobalTransactionMode(direct);
+    attributeTransforms =
+        direct
+            .getRepository()
+            .getTransformations()
+            .values()
+            .stream()
+            .filter(TransformationDescriptor::isTransactional)
```

Comment:
Ok so we can run client side transformations which is just transactional right? Is it correct?

I'm just thinking about that and maybe we should introduce something like transaction scope - client/server. Of course not now, just want to write it somewhere (to be not forgotten) and we can discuss it later. cc @dmvk - I think you have very similar idea

Reply:
That is a complex topic. First - transformations need to be (at least) eventually consistent, that is, when we write the original element (that is input to the transformation), we must make sure, that both the original and the transformed element reach the target commit log. That implies, that we can shift transformation from 'server' to 'client' if and only if, we actually write all the (original and all transformed updates) into *single atomic intermediate commit-log*. That is the case for transactions, where the output of a transaction goes through `_transaction.commit` attribute (which contains all the output of a single transaction) and only then gets replicated to the target commit-logs. That makes it eventually consistent (actually ACID, but that is due how ACID transactions work on top of eventually consistent environment).

Therefore, having a flag for transaction being 'client' or 'server' would mean, that we would have to create intermediate commit-log for 'client' transformation (that are not transactional and therefore have this intermediate commit-log out-of-the-box), and effectively make them 'server'. That makes no practical sense to me.

What is this about is that - if we have guarantee, that transformation are *at least* eventually consistent, we can increase the consistency level to ACID (strongly consistent as long as even pure reads use the correct way - wrap them in a transaction). That is a natural extension, because the writes to transactional attributes are themselves supposed to be ACID, subject to the same conditions about the reads. Therefore I think it is unlikely, that there will _ever_ be any other _client-side_ transformations supported, as the requirement comes not from _where_ the transformation should run but _what guarantees it should give_.